File: src/app/src/main/java/jp/co/e2/dogage/adapter/KindListAdapter.java
package jp.co.e2.dogage.adapter;

import java.util.List;

import jp.co.e2.dogage.R;
import jp.co.e2.dogage.entity.DogMasterEntity;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

/**
 * 種類一覧アダプター
 */
public class KindListAdapter extends ArrayAdapter<DogMasterEntity> {
    private LayoutInflater mInflater;

    /**
     * コンストラクタ
     *
     * @param context コンテキスト
     * @param objects 犬マスタデータ
     */
    public KindListAdapter(Context context, List<DogMasterEntity> objects) {
        super(context, 0, objects);

        if (mInflater == null) {
            mInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        }
    }

    /**
     * ${inheritDoc}
     */
    @Override
    @NonNull
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        // 特定の行(position)のデータを得る
        DogMasterEntity item = getItem(position);

        convertView = mInflater.inflate(R.layout.parts_kind_list, null);

        //頭文字行のラベル
        if (item.getCategory() == null) {
            TextView textViewKind = convertView.findViewById(R.id.textViewKind);
            textViewKind.setVisibility(View.GONE);

            TextView textViewLabel = convertView.findViewById(R.id.textViewLabel);
            textViewLabel.setText(item.getKind());
            textViewLabel.setVisibility(View.VISIBLE);

            isEnabled(position);    //クリックできないように
        }
        //種類
        else {
            TextView textViewKind = convertView.findViewById(R.id.textViewKind);
            textViewKind.setVisibility(View.VISIBLE);
            textViewKind.setText(item.getKind());

            TextView textViewLabel = convertView.findViewById(R.id.textViewLabel);
            textViewLabel.setVisibility(View.GONE);
        }

        return convertView;
    }

    /**
     * ${inheritDoc}
     */
    @Override
    public boolean isEnabled(int position) {
        DogMasterEntity item = getItem(position);

        return (item != null && item.getCategory() != null);
    }
}

File: android/app/src/main/java/jp/co/e2/dogage/adapter/KindListAdapter.kt
package jp.co.e2.dogage.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import jp.co.e2.dogage.R
import jp.co.e2.dogage.entity.DogMasterEntity

/**
 * 種類一覧アダプター
 */
class KindListAdapter(
        private val data: List<DogMasterEntity>,
        private val listener: View.OnClickListener
) : RecyclerView.Adapter<KindListAdapter.ViewHolder>() {

    /**
     * ${inheritDoc}
     */
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflate = LayoutInflater.from(parent.context).inflate(R.layout.parts_kind_list, parent, false)
        val holder = ViewHolder(inflate)

        holder.itemView.setOnClickListener{ view -> listener.onClick(view) }

        return holder
    }

    /**
     * ${inheritDoc}
     */
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = data[position]

        if (item.category == 0) {
            //頭文字行のラベル
            holder.kind.visibility = View.GONE

            holder.label.text = item.kind
            holder.label.visibility = View.VISIBLE

            holder.itemView.isEnabled = false
        } else {
            //種類
            holder.kind.visibility = View.VISIBLE
            holder.kind.text = item.kind

            holder.label.visibility = View.GONE

            holder.itemView.isEnabled = true
        }
    }

    /**
     * ${inheritDoc}
     */
    override fun getItemCount(): Int {
        return data.size
    }

    /**
     * ViewHolder
     */
    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val kind: TextView = view.findViewById<View>(R.id.textViewKind) as TextView
        val label: TextView = view.findViewById(R.id.textViewLabel) as TextView
    }
}